common/convert/convert.go

```diff
@@ -0,0 +1,195 @@
+package convert
+
+import (
+	"encoding/json"
+
+	"github.com/aserto-dev/ds-load/sdk/common/msg"
+	"github.com/aserto-dev/ds-load/sdk/transform"
+	dsc "github.com/aserto-dev/go-directory/aserto/directory/common/v3"
+	"github.com/aserto-dev/scim/common/model"
+	"github.com/elimity-com/scim"
+	"github.com/elimity-com/scim/optional"
+	"github.com/pkg/errors"
+	"github.com/samber/lo"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+var (
+	ErrSourceUserTypeNotSet  = errors.New("source user type not set")
+	ErrSourceGroupTypeNotSet = errors.New("source group type not set")
+	ErrGroupsNotEnabled      = errors.New("groups not enabled")
+)
+
+type Converter struct {
+	cfg *TransformConfig
+}
+
+func NewConverter(cfg *TransformConfig) *Converter {
+	return &Converter{cfg: cfg}
+}
+
+func (c *Converter) ObjectToResource(object *dsc.Object, meta scim.Meta) scim.Resource {
+	eID := optional.String{}
+	attr := c.ObjectToResourceAttributes(object)
+
+	return scim.Resource{
+		ID:         object.GetId(),
+		ExternalID: eID,
+		Attributes: attr,
+		Meta:       meta,
+	}
+}
+
+func (c *Converter) ObjectToResourceAttributes(object *dsc.Object) scim.ResourceAttributes {
+	attr := object.GetProperties().AsMap()
+	delete(attr, "password")
+
+	return attr
+}
+
+func Unmarshal[S any, D any](source S, dest *D) error {
+	data, err := json.Marshal(source)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(data, dest)
+}
+
+func UserToResource(meta scim.Meta, user *model.User) (scim.Resource, error) {
+	attributes := scim.ResourceAttributes{}
+
+	err := Unmarshal(user, &attributes)
+	if err != nil {
+		return scim.Resource{}, err
+	}
+
+	eID := optional.String{}
```

Comment:
:art:
```go
eID := lo.If(user.ExternalID == "", optional.String{}).Else(optional.NewString(user.ExternalID))
```

Reply:
But this may not be the best use-case because all arguments needs to be evaluated in all cases so both optional strings would be constructed.
Just a nifty tool to be aware of.